components/salsa-macros/src/tracked_fn.rs

```diff
@@ -307,6 +325,19 @@ impl syn::visit_mut::VisitMut for ToDbLifetimeVisitor {
     }
 }
 
+struct HasLifetimeVisitor {
+    has_lifetime: bool,
+}
+
+impl<'ast> syn::visit::Visit<'ast> for HasLifetimeVisitor {
+    fn visit_lifetime(&mut self, l: &'ast syn::Lifetime) {
+        // We don't consider `'static` to be a lifetime in this context.
+        if l.ident != "static" {
+            self.has_lifetime = true;
+        }
+    }
+}
```

Comment:
This unfortunately breaks down with macro calls in the type, `-> foo!()` might expand to something introducing an elided lifetime. Likewise a user could use a `Foo<'_>` but without specifying the `<'_>` which warns post 2018 edition, but not error.

Reply:
Both of those seem like reasonable gaps to me, since it will still fallback to the current behaviour. Although I could see the argument that it might create even more confusing behaviour.

What do you think the best path forward is? I have no problems with closing this PR if it doesn't seem worth it.